ADR/0039-install-knative-eventing.md

```diff
@@ -0,0 +1,33 @@
+# 39. Install Knative Eventing
+
+Created: 2024-07-15
+
+## Status
+
+Proposed
```

Comment:
I think that the ADR should describe how Konflux will integrate with Kubearchive. For example how we configure, cluster wide, which resources we want to offload from the cluster what is the api for querying them.

In addition, we need to decide  if it's going to be a hard dependency of Konflux (it means anyone who want to use Konflux will have to install KNative), including users that would like to try out kind on a local cluster such as Kind (for this use case we are striving to make the Konflux instllation as simple as possible).

In addition, all the requirement of Kubearchive should be listed, I believe it also requires a database (maybe other things I'm not aware of).

Reply:
@gbenhaim I advised @skoved to open an ADR _just for getting Knative added to Konflux_ so we could weigh the merits and issues separately from KubeArchive. If you think that is the wrong approach its probably on me.

Reply:
@brianwcook  why should we consider Knative separately from Kubearchive? do you see it (Knative) being used elsewhere in Konflux?

Reply:
> I think that the ADR should describe how Konflux will integrate with Kubearchive. For example how we configure, cluster wide, which resources we want to offload from the cluster what is the api for querying them.
How to configure which resources will be archived cluster wide, is not set in stone yet on the KubeArchive side, but we do have namespace level configuration set

> In addition, we need to decide if it's going to be a hard dependency of Konflux (it means anyone who want to use Konflux will have to install KNative), including users that would like to try out kind on a local cluster such as Kind (for this use case we are striving to make the Konflux instllation as simple as possible).
I'll need to rely on you or other Konflux folks for this part. I'm not super familiar with all of the different components of Konflux and what they need to function correctly. If there are deployments of Konflux where leaving things like `PipelineRuns` on the cluster with using something like Tekton Results to store them off cluster then I would say that Knative Eventing does not need to be hard dependency. **HOWEVER**, KubeArchive will not function at all if Knative Eventing is not install. I also want to be clear that only Knative Eventing needs to be installed. KubeArchive **DOES NOT** depend on any other Knative projects

> In addition, all the requirement of Kubearchive should be listed, I believe it also requires a database (maybe other things I'm not aware of).
KubeArchive needs a PostgreSQL database. For production clusters Konflux will want to set up it's own database ahead of time. I'll add info about how that needs to be done. KubeArchive can be installed with a helm chart, and for non production deployments, that helm chart can optionally install a PostgreSQL database itself. KubeArchive will also need cert-manager installed ahead of time and I can add that to the ADR as well.

Reply:
That is exactly what I want to discuss. Should knative _always_ be deployed? Should we encourage, or at least allow, Knative to be considered in new core feature designs? Or should it be an optional component which is only used when it is dragged in by something like KubeArchive?

Reply:
I want to be clear that KubeArchive depends on Knative Eventing (which can be installed on its own), and **DOES NOT** depend on any other Knative Projects.

If the scope of the ADR should expand to include integration with KubeArchive, I can update this ADR to include all the steps required today for KubeArchive to be installed and configured

> I think that the ADR should describe how Konflux will integrate with Kubearchive. For example how we configure, cluster wide, which resources we want to offload from the cluster what is the api for querying them.

How to configure which resources will be archived cluster wide, is not set in stone yet on the KubeArchive side, but we do have namespace level configuration ironed out and the clusterwide configuration will like be very similar
```
---
# example archiving PipelineRuns in the namespace example
apiVersion: kubearchive.kubearchive.org/v1alpha1
kind: KubeArchiveConfig
metadata:
  name: example-kubearchiveconfig
  namespace: example
spec:
  resources:
    - apiVersion: tekton.dev/v1beta1
      kind: PipelineRun
```

> In addition, we need to decide if it's going to be a hard dependency of Konflux (it means anyone who want to use Konflux will have to install KNative), including users that would like to try out kind on a local cluster such as Kind (for this use case we are striving to make the Konflux instllation as simple as possible).

I'll need to rely on you or other Konflux folks for this part. I'm not super familiar with all of the different components of Konflux and what they need to function correctly. If there are deployments of Konflux where leaving things like `PipelineRuns` on the cluster with using something like Tekton Results to store them off cluster then I would say that Knative Eventing does not need to be hard dependency. **HOWEVER**, KubeArchive will not function at all if Knative Eventing is not install. I also want to be clear that only Knative Eventing needs to be installed. KubeArchive **DOES NOT** depend on any other Knative projects

> In addition, all the requirement of Kubearchive should be listed, I believe it also requires a database (maybe other things I'm not aware of).

KubeArchive needs a PostgreSQL database. For production clusters Konflux will want to set up it's own database ahead of time. KubeArchive can be installed with a helm chart, and for non production deployments, that helm chart can optionally install a PostgreSQL database itself. I'm not sure if Konflux would be interested in this approach for non-production clusters. KubeArchive will also need cert-manager installed ahead of time. If this ADR should expand in scope

Reply:
Even unused features are using resources in dev clusters, konflux is beefy enough with mandatory parts IMO.

Does it bring any benefits or other use cases except being a dependency of kubearchive?

Reply:
>  why should we consider Knative separately from Kubearchive? do you see it (Knative) being used elsewhere in Konflux?

> Does it bring any benefits or other use cases except being a dependency of kubearchive?

I've updated the ADR to include features of Knative Eventing as well as how it could be used by Konflux directly as well

> Should knative always be deployed?

At the very least, Knative **Eventing** must deployed anywhere that Konflux intends to deploy KubeArchive

Reply:
We do have a lot of controllers that look at PipelineRun resources. So it can make architectural sese to use Knative eventing instead of the usual controller watch loops. The watch llops have one major advantage though, they are especially easy to code when you use Kubebuider. In Konflux we are very heavy users of Kubebuilder. When I checked a while ago, I couldn't find anything in the Knative world that comes close to being a comprehensive framework like Kubebuilder. Especially for eventing.

Reply:
I'm not as familiar with the inner workings of konflux as, so thank you for providing more insight.

Knative Eventing, more specifically cloudevents, may not have something like KubeBuilder, but the [cloudevents sdk](https://github.com/cloudevents/sdk-go) provides a really easy way to start receiving events from knative eventing. [This example is extremely simple and just printing out the event id](https://github.com/cloudevents/sdk-go?tab=readme-ov-file#receive-your-first-cloudevent) but i think it demonstrates the point. Also KubeBuilder is performing a lot boilerplate needed for starting the controller and being able to register it with Kubernetes. With Knative Eventing most of the that is offloaded to the configuration of ApiServerSource, to tell it which types of resources to create events for, and Triggers/Subscriptions so that the consumer only receives events that it is interested in. This lets the application consuming the events to only worry about the logic for processing the event and performing some set of required actions based on that event. Because of this difference I think it makes sense why there is not totally comprehensive framework. I do think that the cloudevents sdk provides comprehensive tools that make it incredibly simple to start receiving events so that can focus processing and reacting to them.

I don't think that eventing can completely replace controllers because controllers and eventing are meant to solve different problems. I do think that there are areas that could benefit from Knative Eventing. For example, a controller that is watch tekton PipelineRuns so that when they complete, some action can be performed like sending an email. I think that's an example where using eventing could make more sense. It would receive an event with a PipelineRun object as the payload, it would verify that the PipelineRun completed, grab what ever other data it needs from the object to fill in an email template and then send the email. A controller that has to synchronize the state of different resources on the cluster wouldn't necessarily benefit from something ApiServerSource as easily. However, it could send cloudevents when it performs different actions that other services could subscribe to instead of having to watch resources that this controller synchronizes.